File: src/main/java/no/fint/organization/OrganizationService.java
package no.fint.organization;

import lombok.extern.slf4j.Slf4j;
import no.fint.*;
import no.fint.mailing.MailingService;
import no.fint.model.resource.Link;
import no.fint.model.resource.administrasjon.organisasjon.OrganisasjonselementResource;
import no.fint.model.resource.administrasjon.organisasjon.OrganisasjonselementResources;
import no.fint.utils.ResourceConverter;
import no.fint.utils.RestUtil;
import no.fint.utils.TemplateService;
import org.jooq.lambda.tuple.Tuple;
import org.jooq.lambda.tuple.Tuple2;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

@Service
@Slf4j
public class OrganizationService {

    private final Config config;
    private final RestUtil restUtil;
    private final OrganizationRepository repository;
    private final TemplateService templateService;
    private final MailingService mailingService;

    public OrganizationService(Config config, RestUtil restUtil, OrganizationRepository repository, TemplateService templateService, MailingService mailingService) {
        this.config = config;
        this.restUtil = restUtil;
        this.repository = repository;
        this.templateService = templateService;
        this.mailingService = mailingService;
    }

    @Scheduled(cron = "${fint.orgmonitor.cron}")
    public void update() {
        List<OrganizationDocument> added = new ArrayList<>();
        List<Tuple2<OrganizationDocument, OrganizationDocument>> updated = new ArrayList<>();
        List<OrganizationDocument> updatedDocuments = new ArrayList<>();
        List<String> parentIds = new ArrayList<>();

        List<OrganizationDocument> documents = repository.getAllByOrgId(config.getOrgid());
        log.info("Repository contains {} documents.", documents.size());

        Map<String, OrganizationDocument> organisationMap = documents.stream().collect(Collectors.toMap(r -> r.getData().getOrganisasjonsId().getIdentifikatorverdi(), Function.identity()));


        OrganisasjonselementResources updates = restUtil.getUpdates(new ParameterizedTypeReference<>() {
        }, config.getEndpoint());
        log.info("Found {} updates.", updates.getContent().size());

        for (OrganisasjonselementResource resource : updates.getContent()) {
            String id = resource.getOrganisasjonsId().getIdentifikatorverdi();

            OrganizationDocument current = organisationMap.get(id);

            try {
                if (current == null) {
                    OrganizationDocument document = createDocument(resource);
                    updatedDocuments.add(document);
                    added.add(document);
                    if (StringUtils.hasText(document.getOverordnet())) {
                        OrganizationDocument parent = organisationMap.get(document.overordnetId());
                        if (parent != null && !parentIds.contains(parent.getId())) {
                            parentIds.add(parent.getId());
                        }
                    }
                } else {
                    OrganizationDocument modified = createDocument(resource);
                    if (!modified.equals(current)) {
                        modified.setId(current.getId());
                        updatedDocuments.add(modified);
                        updated.add(Tuple.tuple(current, modified));
                        if (StringUtils.hasText(modified.getOverordnet())) {
                            OrganizationDocument parent = organisationMap.get(modified.overordnetId());
                            if (parent != null && !parentIds.contains(parent.getId())) {
                                parentIds.add(parent.getId());
                            }
                        }
                    }
                }
            } catch (IOException e) {
                log.error("Error converting element", e);
                throw new RuntimeException(e);
            }
        }

        log.info("Saving {} updates ...", updatedDocuments.size());
        repository.saveAll(updatedDocuments);

        log.info("Added: {} items", added.size());

        log.info("Updated: {} items", updated.size());

        if (!added.isEmpty() || !updated.isEmpty()) {
            List<SimpleOrganizationInfo> parentInfo = parentIds.isEmpty() ? new ArrayList<>() : createParentInfo(parentIds);
            String result = templateService.render(added, updated, parentInfo);
            mailingService.send(result);
        }
    }

    private OrganizationDocument createDocument(OrganisasjonselementResource resource) throws IOException {
        OrganizationDocument document = new OrganizationDocument();
        document.setOrgId(config.getOrgid());

        document.setData(ResourceConverter.toOrganisasjonselement(resource));

        document.setOverordnet(
                resource
                        .getOverordnet()
                        .stream()
                        .map(Link::getHref)
                        .findFirst()
                        .orElse(null));

        document.setUnderordnet(
                resource
                        .getUnderordnet()
                        .stream()
                        .map(Link::getHref)
                        .collect(Collectors.toList()));

        return document;
    }

    private List<SimpleOrganizationInfo> createParentInfo(List<String> parentIds) {
            List<SimpleOrganizationInfo> parentInfo = new ArrayList<>();

            for (String id : parentIds) {
                OrganizationDocument document = repository.getOrganizationDocumentByIdAndOrgId(id, config.getOrgid());
                if (document != null && document.getData().getNavn() !=null) {
                    SimpleOrganizationInfo info = new SimpleOrganizationInfo(id, document.getData().getNavn());
                    parentInfo.add(info);
                }
            }
            return parentInfo;
    }
}

File: src/main/kotlin/no/fint/organization/OrganizationService.kt
package no.fint.organization

import no.fint.Config
import no.fint.mailing.MailingService
import no.fint.utils.ResourceConverter
import no.fint.utils.RestUtil
import no.fint.utils.TemplateService
import no.novari.fint.model.resource.administrasjon.organisasjon.OrganisasjonselementResource
import no.novari.fint.model.resource.administrasjon.organisasjon.OrganisasjonselementResources
import org.slf4j.LoggerFactory
import org.springframework.core.ParameterizedTypeReference
import org.springframework.stereotype.Service
import org.springframework.util.StringUtils

/**
 * Service responsible for monitoring and updating organization elements.
 *
 * This service fetches organization data from a remote endpoint, compares it with the local database,
 * and identifies added or modified elements. If changes are detected, it updates the database
 * and triggers an email notification with a summary of the changes.
 *
 */
@Service
class OrganizationService(
    private val config: Config,
    private val organizationRepository: OrganizationRepository,
    private val restUtil: RestUtil,
    private val mailingService: MailingService,
    private val templateService: TemplateService,
) {
    private val logger = LoggerFactory.getLogger(javaClass)

    private data class UpdateChanges(
        val addedDocuments: MutableList<OrganizationDocument> = mutableListOf(),
        val documentsToSave: MutableList<OrganizationDocument> = mutableListOf(),
        val updatedPairs: MutableList<Pair<OrganizationDocument, OrganizationDocument>> = mutableListOf(),
        val parentIds: MutableSet<String> = linkedSetOf(),
    ) {
        fun hasChanges(): Boolean = addedDocuments.isNotEmpty() || updatedPairs.isNotEmpty()
    }

    /**
     * Orchestrates the update process for organization elements.
     *
     * This method performs the following steps:
     * 1. Retrieves all existing organization documents from the repository.
     * 2. Fetches the latest updates from the configured REST endpoint.
     * 3. Compares the fetched resources with existing documents to detect additions and modifications.
     * 4. Persists the changes (new and updated documents) to the database.
     * 5. If changes were found, compiles a report (including parent info) and sends it via email.
     */
    fun update() {
        val documents = organizationRepository.getAllByOrgId(config.orgid)
        logger.info("Repository contains ${documents.size} documents for ${config.orgid}.")
        val organizationMap = documents.associateBy { it.data?.organisasjonsId?.identifikatorverdi }

        val updates = fetchUpdates() ?: return
        logger.info("Found ${updates.size} updates")
        logger.trace(
            "Updates content: {}",
            updates.content.joinToString(", ") { "ID=${it.organisasjonsId.identifikatorverdi}, Name=${it.navn}" },
        )
        val changes = collectChanges(updates.content, organizationMap)

        logger.info("Saving {} updates", changes.documentsToSave.size)
        organizationRepository.saveAll(changes.documentsToSave)
        logger.info("Added: {} items", changes.addedDocuments.size)
        logger.info("Updated: {} items", changes.updatedPairs.size)

        sendNotification(changes)
    }

    private fun fetchUpdates(): OrganisasjonselementResources? {
        val updates =
            restUtil.getUpdates(
                object : ParameterizedTypeReference<OrganisasjonselementResources?>() {},
            )

        if (updates == null) {
            logger.error("Failed to fetch updates from endpoint")
        }

        return updates
    }

    private fun collectChanges(
        resources: List<OrganisasjonselementResource>,
        organizationMap: Map<String?, OrganizationDocument>,
    ): UpdateChanges {
        val changes = UpdateChanges()

        resources.forEach { resource ->
            val resourceId = resource.organisasjonsId.identifikatorverdi
            val newDocument = createDocument(resource)
            val existingDocument = organizationMap[resourceId]

            if (existingDocument == null) { // New document
                storeNewDocument(newDocument, changes, organizationMap)
            } else if (existingDocument != newDocument) { // Changed document
                storeModifiedDocument(newDocument, resourceId, existingDocument, changes, organizationMap)
            }
            // If none of the above, then we do nothing as there is no change.
        }

        return changes
    }

    private fun storeModifiedDocument(
        modifiedDocument: OrganizationDocument,
        resourceId: String,
        existingDocument: OrganizationDocument,
        changes: UpdateChanges,
        organizationMap: Map<String?, OrganizationDocument>,
    ) {
        modifiedDocument.id = resourceId
        changes.documentsToSave.add(modifiedDocument)

        // A little ugly to use .copy(), but necessary to avoid existingDocument being modified after it is added to updatedPairs
        // The pointer to underordnet is copied.
        changes.updatedPairs.add(
            Pair(
                existingDocument.copy(underordnet = existingDocument.underordnet?.toList()),
                modifiedDocument,
            ),
        )

        registerParentId(modifiedDocument.overordnet, organizationMap, changes)
    }

    private fun storeNewDocument(
        newDocument: OrganizationDocument,
        changes: UpdateChanges,
        organizationMap: Map<String?, OrganizationDocument>,
    ) {
        changes.documentsToSave.add(newDocument)
        changes.addedDocuments.add(newDocument)

        registerParentId(newDocument.overordnet, organizationMap, changes)
    }

    private fun registerParentId(
        parentHref: String?,
        organizationMap: Map<String?, OrganizationDocument>,
        changes: UpdateChanges,
    ) {
        if (!StringUtils.hasText(parentHref)) {
            return
        }

        organizationMap[parentHref]?.id?.let(changes.parentIds::add)
    }

    private fun sendNotification(changes: UpdateChanges) {
        if (!changes.hasChanges()) {
            return
        }

        val parentInfo = createParentInfo(changes.parentIds.toList())
        mailingService.send(templateService.render(changes.addedDocuments, changes.updatedPairs, parentInfo))
    }

    private fun createDocument(resource: OrganisasjonselementResource) =
        OrganizationDocument().apply {
            orgId = config.orgid
            data = ResourceConverter.toOrganisasjonselement(resource)
            overordnet = resource.overordnet.firstOrNull()?.href
            underordnet = resource.underordnet.map { it.href }
        }

    private fun createParentInfo(parentIds: List<String>): List<SimpleOrganizationInfo> =
        parentIds.mapNotNull { id ->
            organizationRepository.getOrganizationDocumentByIdAndOrgId(id, config.orgid)?.let { document ->
                document.data?.navn?.let { navn ->
                    SimpleOrganizationInfo(id, navn)
                }
            }
        }
}
